### ISSUE TITLE
CI: Docstring validation is slow

### ISSUE THREAD

**Issue by lithomas1** [pandas-dev/pandas#57578]
It takes around 22 minutes, for some reason.

**Comment by mroeschke** [comment]
> tbh I think we should rip all the docstring sharing logic, just have plain old docstrings without substitutions,

Big +1

**Comment by MarcoGorelli** [comment]
I started this here: https://github.com/pandas-dev/pandas/pull/57683

I have a little script to do this, which seems to working well enough. Before I go too far, just checking people are OK with it

Pros of keeping docstring substitutions:
- shorter docstrings within the code
- helps dedupe some parameter definitions

Pros of removing docstring substitions:
- drastically faster docstring checking thanks to ruff
- blazingly fast docstring formatting becomes possible
- easier to read docstrings (personally, with the substitutions, I only have confidence that they're correct if I see a screenshot or use `/preview`)
- unblocks using `darglint` rules (which may make their way into ruff!) https://github.com/astral-sh/ruff/issues/458

I don't think ruff replaces _all_ of the `validate_docstrings` script, but that script can be used with or without docstring substitutions. But removing docstring substitutions at least means we can start removing things from `validate_docstrings` which are covered by ruff and cut down on CI like this

**Comment by dontgoto** [comment]
Wasn't aware of this issue and wrote #57826 after I introduced a docstring error in my first PR and got annoyed at the speed of the docstring checks. It reduces the runtime of the `check_code.sh docstrings` down to 2-3 minutes again.

I also have another PR ready to go once #57826 is merged. That second PR brings the validation down to about 10 seconds. Main issue was the way the temp files were created and how pep8 was called via the shell.

**Comment by mroeschke** [comment]
> So maybe with a 10 second runtime it's not necessary to do any functionality changes to the docstring validation and sharing in the end?

I think the docstring sharing is still worth removing to:

1. Make it easier for new contributors to work on docstrings due to less complexity
2. Allow `ruff` to take over docstring formatting and validation

**Comment by datapythonista** [comment]
I'm happy to move towards removing or reducing the docstring sharing. But based on my experience getting fully rid of it may take years. If we want to proactively work on duplicating now reused docstrings, I'd start by the ones with higher complexity and see how that goes, before opening issues to remove all them.
